cpp/src/arrow/type_traits.cc

```
@@ -0,0 +1,94 @@
+                                   // check before casting to an array of structs
+    case Type::BOOL:               // Always treated as uint8_t*
+    case Type::INT8:               // Always treated as uint8_t*
+    case Type::UINT8:              // Always treated as uint8_t*
+    case Type::SPARSE_UNION:       // No second buffer
+    case Type::RUN_END_ENCODED:    // No buffers
+    case Type::STRUCT:             // No second buffer
+      return 1;
+    case Type::INT16:
+    case Type::UINT16:
+    case Type::HALF_FLOAT:
+      return 2;
+    case Type::INT32:
+    case Type::UINT32:
+    case Type::FLOAT:
+    case Type::STRING:  // Offsets may be cast to int32_t*
+    case Type::BINARY:  // Offsets may be cast to int32_t*
+    case Type::DATE32:
+    case Type::TIME32:
+    case Type::LIST:         // Offsets may be cast to int32_t*, data is in child array
+    case Type::MAP:          // This is a list array
+    case Type::DENSE_UNION:  // Has an offsets buffer of int32_t*
+    case Type::INTERVAL_MONTHS:    // Stored as int32_t*
+    case Type::INTERVAL_DAY_TIME:  // Stored as two contiguous 32-bit integers
+      return 4;
+    case Type::INT64:
+    case Type::UINT64:
+    case Type::DOUBLE:
+    case Type::DECIMAL128:    // May be cast to GenericBasicDecimal* which requires
+                              // alignment of 8
+    case Type::DECIMAL256:    // May be cast to GenericBasicDecimal* which requires
+                              // alignment of 8
+    case Type::LARGE_BINARY:  // Offsets may be cast to int64_t*
+    case Type::LARGE_LIST:    // Offsets may be cast to int64_t*
+    case Type::LARGE_STRING:  // Offsets may be cast to int64_t*
+    case Type::DATE64:
+    case Type::TIME64:
+    case Type::TIMESTAMP:
+    case Type::DURATION:
+    case Type::INTERVAL_MONTH_DAY_NANO:  // Stored as two 32-bit integers and a 64-bit
+                                         // integer
+      return 8;
+    default:
```

Can we keep this without a `default` so when new types are added by me and @bkietz we get a compiler error reminding us to think about this function?

---

Thanks for pointing this out.  I thought a switch case without a default just silently ignored options but I was wrong.

---

I've removed the default (and manually listed the remaining invalid options)